snapcraft/internal/pluginhandler/__init__.py

```diff
@@ -571,13 +571,25 @@ def prime(self, force=False) -> None:  # noqa: C901
         elif is_classic:
             dynamic_linker = self._project_options.get_core_dynamic_linker()
 
-        if dynamic_linker:
+        if dynamic_linker and not self._build_attributes.no_patchelf():
+            logger.warning(
+                'Files in this part are going to be patched to work correctly '
+                'in the environment.\n'
+                'To disable this behavior set '
+                '`build-properties: [no-patchelf]` for the part.')
             elf_patcher = elf.Patcher(
                 dynamic_linker=dynamic_linker,
                 root_path=self.primedir,
                 preferred_patchelf_path=staged_patchelf_path)
             for elf_file in elf_files:
                 elf_patcher.patch(elf_file=elf_file)
+        elif dynamic_linker and self._build_attributes.no_patchelf():
+            logger.warning(
+                'The following files are not going to be patched to work '
+                'correctly in the environment as '
+                '`build-attributes: [no-patchelf]` is set for the '
+                'part:\n{}'.format(
```

Comment:
I suggest:
The following files may not execute correctly in all environments because they won't be patched (build-attributes: [no-patchelf] is set for the part.

Reply:
Note that you also have multiple placeholders, but only a single replacement.

Reply:
Also, the list printed here could be massive. Even for the tiny opencv-demo, it looks like this:

```
The following files are not going to be patched to work correctly in the environment as `build-attributes: [no-patchelf]` is set for the part:
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libtbb.so.2- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/r200_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_intel.so.1.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-dri2.so.0.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/lib/x86_64-linux-gnu/libexpat.so.1.6.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/nouveau_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-glx.so.0.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_radeon.so.1.0.1- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1.2.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-sync.so.1.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/radeon_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-dri3.so.0.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libglapi.so.0.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/swrast_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXxf86vm.so.1.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/lib/x86_64-linux-gnu/libbsd.so.0.8.2- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/bin/example- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libtbbmalloc_proxy.so.2- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/i915_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXext.so.6.4.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/i965_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb.so.1.1.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_amdgpu.so.1.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/virtio_gpu_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libX11-xcb.so.1.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libelf-0.165.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/r300_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/nouveau_vieux_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/radeonsi_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libX11.so.6.3.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libexpatw.so.1.6.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libedit.so.2.0.53- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXdamage.so.1.1.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXfixes.so.3.1.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libpciaccess.so.0.11.1- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/kms_swrast_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXau.so.6.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXdmcp.so.6.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libsensors.so.4.4.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libtbbmalloc.so.2- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/vmwgfx_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_nouveau.so.2.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-present.so.0.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libffi.so.6.0.4- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/r600_dri.so- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libopencv_core.so.2.4.9- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxshmfence.so.1.0.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm.so.2.4.0- {}
/home/kyrofa/src/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libLLVM-5.0.so.1
```

Reply:
fixed output

```
Priming example 
The following files are not going to be patched to work correctly in the environment as `build-attributes: [no-patchelf]` is set for the part:
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/nouveau_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/r600_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/i915_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXau.so.6.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_intel.so.1.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libffi.so.6.0.4
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/r200_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-dri2.so.0.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libX11-xcb.so.1.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libLLVM-5.0.so.1
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/vmwgfx_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libglapi.so.0.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libtbbmalloc.so.2
- /home/ubuntu/source/snapcraft/demos/opencv/prime/lib/x86_64-linux-gnu/libbsd.so.0.8.2
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-glx.so.0.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libedit.so.2.0.53
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libpciaccess.so.0.11.1
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXdamage.so.1.1.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXfixes.so.3.1.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/lib/x86_64-linux-gnu/libexpat.so.1.6.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/kms_swrast_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/r300_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libX11.so.6.3.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb.so.1.1.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_radeon.so.1.0.1
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/i965_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/radeonsi_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/bin/example
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-dri3.so.0.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libtbbmalloc_proxy.so.2
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/nouveau_vieux_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libexpatw.so.1.6.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-present.so.0.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXxf86vm.so.1.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libelf-0.165.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/virtio_gpu_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_amdgpu.so.1.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/radeon_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/dri/swrast_dri.so
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1.2.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libsensors.so.4.4.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxcb-sync.so.1.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libxshmfence.so.1.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libtbb.so.2
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libopencv_core.so.2.4.9
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXdmcp.so.6.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm_nouveau.so.2.0.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libXext.so.6.4.0
- /home/ubuntu/source/snapcraft/demos/opencv/prime/usr/lib/x86_64-linux-gnu/libdrm.so.2.4.0
```

It is long, but still valid

Reply:
This output will be smaller with #1959